typing/typecore.ml

```diff
@@ -6824,6 +6832,44 @@ let escaping_hint reason (context : Env.escaping_context option) =
       [ Location.msg
           "@[Hint: It is captured by a partial application@]" ]
   | _, _ -> []
+  end
+  @
+  begin match submode_reason with
+  | Application result_ty ->
+    (* [get_non_local_arity ty] returns [Some (n_args, sureness)] iff [ty] is a
+       function type with [n_args] arguments and its return type is
+       local. [sureness] <=> the return type is definitely local. *)
+    let get_non_local_arity ty =
+      let rec loop alloc n ty =
+        match get_desc ty with
+        | Tarrow ((_, _, res_mode), _, res_ty, _) ->
+          begin match Alloc_mode.check_const res_mode with
+          | Some Global -> Some (n+1, true)
+          | _ -> loop res_mode (n+1) res_ty
+          end
+        | _ ->
+          if n = 0
+          then None
+          else begin match Alloc_mode.check_const alloc with
+            | Some Local -> None
+            | Some Global -> assert false (* this should have been caught above *)
+            | None -> Some (n, false)
+          end
+      in
+      loop (Amode Local) (* doesn't matter *)
+           0 ty
```

Comment:
I found the `alloc` parameter logic in `loop` a little fiddly, so here's a suggestion for an alternative:
```ocaml
let rec loop sureness n ty =
  match get_desc ty with
  | Tarrow ((_, _, res_mode), _, res_ty, _) ->
    begin match Alloc_mode.check_const res_mode with
    | Some Global ->
      Some (n+1, true)
    | (None | Some Local) as res_mode ->
      let sureness = sureness && Option.is_some res_mode in
      loop sureness (n+1) res_ty
    end
  | _ ->
    if n = 0
    then None
    else Some (n, sureness)
in
loop true 0 ty
```

Reply:
Perfect. Thanks!